Source/Dafny/Resolver.cs

```diff
@@ -11100,6 +11101,21 @@ public class ResolveTypeReturn {
                 var call = new CallStmt(methodCallInfo.Tok, s.EndTok, new List<Expression>(), methodCallInfo.Callee, methodCallInfo.ActualParameters);
                 s.ResolvedStatements.Add(call);
               }
+            } else if (expr is NameSegment or ExprDotName) {
+              if (expr is NameSegment) {
+                ResolveNameSegment(expr as NameSegment, true, null, revealResolutionContext, true);
+              } else {
+                ResolveDotSuffix(expr as ExprDotName, true, null, revealResolutionContext, true);
```

Comment:
For `expr as NameSegment` and `expr as ExprDotName`, you expect that `expr` will have those respective types. In other words, you expect that the `as` will not return `null`. To indicate this in the program text, I suggest you instead use a cast: `(NameSegment)expr` and `(ExprDotName)expr`.